File: occasum-core/src/main/java/pw/stamina/occasum/registry/RegisterWithQuery.java
package pw.stamina.occasum.registry;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.PropertyNode;
import pw.stamina.occasum.scan.PropertyScanningStrategy;
import pw.stamina.occasum.scan.result.PropertyScanResult;

import java.util.function.Function;

public interface RegisterWithQuery {

    PropertyScanResult getResult();

    Function<PropertyRegistry, PropertyNode> getNodeExtractor();

    static Builder builder(PropertyHandle handle) {
        return null;
    }

    interface Builder {

        Builder handle(PropertyHandle handle);

        Builder source(Object source);

        Builder handleAsSource();

        Builder registerToHandleRoot();

        Builder registerToRootOf(PropertyHandle handle);

        Builder registerToNode(PropertyNode node);

        Builder scanningStrategy(PropertyScanningStrategy scanningStrategy);

        RegisterWithQuery build();
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/registry/RegisterWithQuery.kt
package pw.stamina.occasum.registry

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.PropertyNode
import pw.stamina.occasum.scan.PropertyScanningStrategy
import pw.stamina.occasum.scan.result.PropertyScanResult

interface RegisterWithQuery {

    val result: PropertyScanResult

    val nodeExtractor: (PropertyRegistry) -> PropertyNode

    interface Builder {

        fun handle(handle: PropertyHandle): Builder

        fun source(source: Any): Builder

        fun handleAsSource(): Builder

        fun registerToHandleRoot(): Builder

        fun registerToRootOf(handle: PropertyHandle): Builder

        fun registerToNode(node: PropertyNode): Builder

        fun scanningStrategy(scanningStrategy: PropertyScanningStrategy): Builder

        fun build(): RegisterWithQuery
    }

    companion object {

        fun builder(handle: PropertyHandle): Builder? {
            return null
        }
    }
}
